### PR TITLE
add security group for zscaler 

### PR DESCRIPTION
**Pull request by knavapbc**
## 🎫 Ticket

https://jira.cms.gov/browse/PLT-471

## 🛠 Changes

Add security group and security group rule on all the alb so clients on zscaler can access.

### CODE REVIEW

File: terraform/services/zscaler-security-groups/variables.tf [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+variable "vpc_id_public" {
+  description = "The vpc id public"
+  type        = string
+}
+
+variable "vpc_id_private" {
+  description = "The vpc id private"
+  type        = string
+}
```
Comment (gsf) [inline terraform/services/zscaler-security-groups/variables.tf:27]: The vpc can be determined by the vpc module in main.tf. See how this is done at https://github.com/CMSgov/ab2d-bcda-dpc-platform/blob/0c40661080b9f0b49eef82237f6dbafecfc244b8/terraform/modules/function/main.tf#L181

Comment (knavapbc) [reply terraform/services/zscaler-security-groups/variables.tf:22]: Wait, i don't think that's how it works. That's a vpc module, but we aren't creating a vpc we are referencing an existing one. So it's either an output.tf, from a remote state file , or an actual input.

File: terraform/services/zscaler-security-groups/README.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# Terraform for zscaler related resources
+
+Creating public and private security group for loadbalancers
```
[the remark this answers is not in the document]
Comment (gsf) [reply terraform/services/zscaler-security-groups/README.md:3]: Why was this change not integrated?

File: terraform/services/zscaler-security-groups/README.md [hunk 3]
```diff
@@ -0,0 +1 @@
+# Terraform for zscaler related resources
```
[the remark this answers is not in the document]
Comment (gsf) [reply terraform/services/zscaler-security-groups/README.md:1]: Why was this change not integrated?

File: terraform/services/zscaler-security-groups/main.tf [hunk 4]
```diff
@@ -0,0 +1,23 @@
+resource "aws_security_group" "zscaler_traffic_allow_public" {
+  name        = "${var.app}-${var.env}-allow-zscaler-public"
+  description = "Allow internet zscaler traffic public"
+  vpc_id      = var.vpc_id
+  ingress {
+    from_port   = 0
+    to_port     = 0
+    protocol    = -1
+    cidr_blocks = ["65.213.206.0/24","208.250.57.0/24"]
```
Comment (gsf) [inline terraform/services/zscaler-security-groups/main.tf:9]: Put CIDRs on separate lines for readability and easier updating

File: terraform/services/zscaler-security-groups/variables.tf (hunk 1 above)
Comment (gsf) [reply terraform/services/zscaler-security-groups/variables.tf:22]: The vpc module gets us a reference to the existing VPC

File: terraform/services/zscaler-security-groups/main.tf [hunk 5]
```diff
@@ -0,0 +1,23 @@
+resource "aws_security_group" "zscaler_traffic_allow_public" {
```
Comment (gsf) [inline terraform/services/zscaler-security-groups/main.tf:1]: "traffic_allow" is unnecessary as that's what a security group generally does

```suggestion
resource "aws_security_group" "zscaler_public" {
```

File: terraform/services/zscaler-security-groups/main.tf [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  name        = "${var.app}-${var.env}-allow-zscaler-public"
+  description = "Allow internet zscaler traffic public"
```
Comment (gsf) [inline terraform/services/zscaler-security-groups/main.tf:3]:
```suggestion
  description = "Allow public zscaler traffic"
```

File: terraform/services/zscaler-security-groups/main.tf [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+  vpc_id      = var.vpc_id
+  ingress {
```
Comment (gsf) [inline terraform/services/zscaler-security-groups/main.tf:5]: The way security groups are managed in the AWS terraform provider has recently been updated. These CIDRs should be moved to aws_vpc_security_group_ingress_rule resources. See the first note at https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/security_group.html:

> The [aws_vpc_security_group_egress_rule](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/vpc_security_group_egress_rule) and [aws_vpc_security_group_ingress_rule](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/vpc_security_group_ingress_rule) resources have been added to address these limitations and should be used for all new security group rules.

**Comment by knavapbc** [comment]
```
var.app
  The application name (ab2d, bcda, dpc)

  Enter a value: ab2d

var.env
  The application environment (dev, test, mgmt, sbx, prod)

  Enter a value: dev

module.vpc.data.aws_vpc.this: Reading...
module.vpc.data.aws_vpc.this: Read complete after 1s [id=vpc-0c6413ec40c5fdac3]

Terraform used the selected providers to generate the following execution plan. Resource actions are
indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # aws_security_group.zscaler_private will be created
  + resource "aws_security_group" "zscaler_private" {
      + arn                    = (known after apply)
      + description            = "Allow internet zscaler traffic private"
      + egress                 = (known after apply)
      + id                     = (known after apply)
      + ingress                = (known after apply)
      + name                   = "ab2d-dev-allow-zscaler-private"
      + name_prefix            = (known after apply)
      + owner_id               = (known after apply)
      + revoke_rules_on_delete = false
      + tags_all               = {
          + "business"  = "oeda"
          + "code"      = "https://github.com/CMSgov/ab2d-bcda-dpc-platform/tree/main/terraform/services/zscaler-security-groups"
          + "terraform" = "true"
        }
      + vpc_id                 = "vpc-0c6413ec40c5fdac3"
    }

  # aws_security_group.zscaler_public will be created
  + resource "aws_security_group" "zscaler_public" {
      + arn                    = (known after apply)
      + description            = "Allow public zscaler traffic"
      + egress                 = (known after apply)
      + id                     = (known after apply)
      + ingress                = (known after apply)
      + name                   = "ab2d-dev-allow-zscaler-public"
      + name_prefix            = (known after apply)
      + owner_id               = (known after apply)
      + revoke_rules_on_delete = false
      + tags_all               = {
          + "business"  = "oeda"
          + "code"      = "https://github.com/CMSgov/ab2d-bcda-dpc-platform/tree/main/terraform/services/zscaler-security-groups"
          + "terraform" = "true"
        }
      + vpc_id                 = "vpc-0c6413ec40c5fdac3"
    }

  # aws_security_group_rule.zscaler_allow_private will be created
  + resource "aws_security_group_rule" "zscaler_allow_private" {
      + cidr_blocks              = [
          + "10.223.126.64/27",
          + "10.223.126.96/28",
          + "10.220.126.64/27",
          + "10.220.126.96/28",
        ]
      + from_port                = 0
      + id                       = (known after apply)
      + protocol                 = "-1"
      + security_group_id        = (known after apply)
      + security_group_rule_id   = (known after apply)
      + self                     = false
      + source_security_group_id = (known after apply)
      + to_port                  = 0
      + type                     = "ingress"
    }

  # aws_security_group_rule.zscaler_allow_public will be created
  + resource "aws_security_group_rule" "zscaler_allow_public" {
      + cidr_blocks              = [
          + "65.213.206.0/24",
          + "208.250.57.0/24",
        ]
      + from_port                = 0
      + id                       = (known after apply)
      + protocol                 = "-1"
      + security_group_id        = (known after apply)
      + security_group_rule_id   = (known after apply)
      + self                     = false
      + source_security_group_id = (known after apply)
      + to_port                  = 0
      + type                     = "ingress"
    }

Plan: 4 to add, 0 to change, 0 to destroy.


```

File: terraform/services/zscaler-security-groups/main.tf (hunk 7 above)
Comment (gsf) [reply terraform/services/zscaler-security-groups/main.tf]: Please update to use aws_vpc_security_group_ingress_rule resources as noted in the comment above.

**Comment by knavapbc** [comment]
```

Plan: 8 to add, 0 to change, 0 to destroy.
```

**Comment by knavapbc** [comment]
@gsf

File: terraform/services/zscaler-security-groups/main.tf [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  description = "List of names"
+  default = [
+    "10.223.126.64/27",
+    "10.223.126.96/28",
+    "10.220.126.64/27",
+    "10.220.126.96/28"
+  ]
+}
+resource "aws_vpc_security_group_ingress_rule" "zscaler_allow_private" {
+  for_each          = toset(var.prviate_list)
```
Comment (gsf) [inline terraform/services/zscaler-security-groups/main.tf:47]: I think this `for_each` is the way to go. Do this for the public group as well and add `private_cidrs` and `public_cidrs` to variables.tf without a default. We can then pass those variables in in the workflow with aws-params-env-action as at https://github.com/CMSgov/ab2d-bcda-dpc-platform/blob/f394bd501b6ac9f0f65248532602ee06960de7cf/.github/workflows/build-runner-images.yml#L33.

File: terraform/services/zscaler-security-groups/main.tf [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+  from_port         = 0
+  to_port           = 0
```
Comment (gsf) [inline terraform/services/zscaler-security-groups/main.tf:49]: No from_port or to_port on `ip_protocol = -1` - see docs on aws_vpc_security_group_ingress_rule.

File: terraform/services/zscaler-security-groups/main.tf (hunk 8 above)
Comment (gsf) [reply terraform/services/zscaler-security-groups/main.tf:29]: This is looking good. Please set up the plan and apply workflows so we can check the output.

File: terraform/services/zscaler-security-groups/variables.tf [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+
+variable "private_list" {
+  type        = list(string)
+  description = "private ip list"
+}
+variable "public_list" {
+  type        = list(string)
+  description = "public ip list"
+}
```
Comment (gsf) [inline terraform/services/zscaler-security-groups/variables.tf:27]: As noted in my other comment, let's use `private_cidrs` and `public_cidrs` to be more specific about what's expected here. Also, if we're passing these in as strings from AWS parameter store we likely won't be able to use the `list(string)` type.

And a note on formatting: use one blank line between each variable block.

**Comment by gsf** [comment]
On thinking about this further, I realize we're facing a similar issue to the one we're working on in [PLT-348](https://jira.cms.gov/browse/PLT-348) regarding lists of IP ranges. In order to manage this in the same way, let's leave the security group rules out of the terraform. We'll set them manually for now and move toward automation in a future ticket.

**Comment by knavapbc** [comment]
```

hk@HKs-MacBook-Pro zscaler-security-groups % terraform plan
var.app
  The application name (ab2d, bcda, dpc)

  Enter a value: ab2d

var.env
  The application environment (dev, test, mgmt, sbx, prod)

  Enter a value: dev

var.private_cidrs
  private ip list

  Enter a value: ["192.168.0.0/24","10.0.0.0/24"]

var.public_cidrs
  public ip list

  Enter a value: ["192.168.1.0/24","10.1.0.0/24"]

module.vpc.data.aws_vpc.this: Reading...
module.vpc.data.aws_vpc.this: Read complete after 1s [id=vpc-0c6413ec40c5fdac3]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the
following symbols:
  + create

Terraform will perform the following actions:

  # aws_security_group.zscaler_private will be created
  + resource "aws_security_group" "zscaler_private" {
      + arn                    = (known after apply)
      + description            = "Allow internet zscaler traffic private"
      + egress                 = (known after apply)
      + id                     = (known after apply)
      + ingress                = (known after apply)
      + name                   = "ab2d-dev-allow-zscaler-private"
      + name_prefix            = (known after apply)
      + owner_id               = (known after apply)
      + revoke_rules_on_delete = false
      + tags_all               = {
          + "business"  = "oeda"
          + "code"      = "https://github.com/CMSgov/ab2d-bcda-dpc-platform/tree/main/terraform/services/zscaler-security-groups"
          + "terraform" = "true"
        }
      + vpc_id                 = "vpc-0c6413ec40c5fdac3"
    }

  # aws_security_group.zscaler_public will be created
  + resource "aws_security_group" "zscaler_public" {
      + arn                    = (known after apply)
      + description            = "Allow public zscaler traffic"
      + egress                 = (known after apply)
      + id                     = (known after apply)
      + ingress                = (known after apply)
      + name                   = "ab2d-dev-allow-zscaler-public"
      + name_prefix            = (known after apply)
      + owner_id               = (known after apply)
      + revoke_rules_on_delete = false
      + tags_all               = {
          + "business"  = "oeda"
          + "code"      = "https://github.com/CMSgov/ab2d-bcda-dpc-platform/tree/main/terraform/services/zscaler-security-groups"
          + "terraform" = "true"
        }
      + vpc_id                 = "vpc-0c6413ec40c5fdac3"
    }

  # aws_vpc_security_group_ingress_rule.zscaler_allow_private["10.0.0.0/24"] will be created
  + resource "aws_vpc_security_group_ingress_rule" "zscaler_allow_private" {
      + arn                    = (known after apply)
      + cidr_ipv4              = "10.0.0.0/24"
      + id                     = (known after apply)
      + ip_protocol            = "-1"
      + security_group_id      = (known after apply)
      + security_group_rule_id = (known after apply)
      + tags_all               = {
          + "business"  = "oeda"
          + "code"      = "https://github.com/CMSgov/ab2d-bcda-dpc-platform/tree/main/terraform/services/zscaler-security-groups"
          + "terraform" = "true"
        }
    }

  # aws_vpc_security_group_ingress_rule.zscaler_allow_private["192.168.0.0/24"] will be created
  + resource "aws_vpc_security_group_ingress_rule" "zscaler_allow_private" {
      + arn                    = (known after apply)
      + cidr_ipv4              = "192.168.0.0/24"
      + id                     = (known after apply)
      + ip_protocol            = "-1"
      + security_group_id      = (known after apply)
      + security_group_rule_id = (known after apply)
      + tags_all               = {
          + "business"  = "oeda"
          + "code"      = "https://github.com/CMSgov/ab2d-bcda-dpc-platform/tree/main/terraform/services/zscaler-security-groups"
          + "terraform" = "true"
        }
    }

  # aws_vpc_security_group_ingress_rule.zscaler_allow_public["10.1.0.0/24"] will be created
  + resource "aws_vpc_security_group_ingress_rule" "zscaler_allow_public" {
      + arn                    = (known after apply)
      + cidr_ipv4              = "10.1.0.0/24"
      + id                     = (known after apply)
      + ip_protocol            = "-1"
      + security_group_id      = (known after apply)
      + security_group_rule_id = (known after apply)
      + tags_all               = {
          + "business"  = "oeda"
          + "code"      = "https://github.com/CMSgov/ab2d-bcda-dpc-platform/tree/main/terraform/services/zscaler-security-groups"
          + "terraform" = "true"
        }
    }

  # aws_vpc_security_group_ingress_rule.zscaler_allow_public["192.168.1.0/24"] will be created
  + resource "aws_vpc_security_group_ingress_rule" "zscaler_allow_public" {
      + arn                    = (known after apply)
      + cidr_ipv4              = "192.168.1.0/24"
      + id                     = (known after apply)
      + ip_protocol            = "-1"
      + security_group_id      = (known after apply)
      + security_group_rule_id = (known after apply)
      + tags_all               = {
          + "business"  = "oeda"
          + "code"      = "https://github.com/CMSgov/ab2d-bcda-dpc-platform/tree/main/terraform/services/zscaler-security-groups"
          + "terraform" = "true"
        }
    }

Plan: 6 to add, 0 to change, 0 to destroy.

────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Note: You didn't use the -out option to save this plan, so Terraform can't guarantee to take exactly these actions if you
run "terraform apply" now.
Releasing state lock. This may take a few moments...

```

File: terraform/modules/vpc/variables.tf [hunk 11]
```diff
@@ -11,7 +11,7 @@ variable "env" {
   description = "The application environment (dev, test, sbx, prod)"
   type        = string
   validation {
-    condition     = contains(["dev", "test", "sbx", "prod"], var.env)
+    condition     = contains(["dev", "test", "sbx", "prod", "mgmt"], var.env)
     error_message = "Valid value for env is dev, test, sbx, or prod."
```
Comment (gsf) [inline terraform/modules/vpc/variables.tf:15]:
```suggestion
    error_message = "Valid value for env is dev, test, sbx, prod, or mgmt."
```

File: terraform/services/zscaler-security-groups/main.tf [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+  description = "Allow internet zscaler traffic private"
+  vpc_id      = module.vpc.id
+}
+
+#resource "aws_vpc_security_group_ingress_rule" "zscaler_allow_private" {
+#  for_each          = toset(var.private_cidrs)
+#  ip_protocol       = -1
+#  cidr_ipv4         = each.key
+#  security_group_id = aws_security_group.zscaler_private.id
+#}
```
Comment (gsf) [inline terraform/services/zscaler-security-groups/main.tf:33]: Drop these commented blocks

File: terraform/services/zscaler-security-groups/variables.tf [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+
+
+#variable "private_cidrs" {
+#  type        = list(string)
+#  description = "private ip list"
+#}
+#variable "public_cidrs" {
+#  type        = list(string)
+#  description = "public ip list"
+#}
```
Comment (gsf) [inline terraform/services/zscaler-security-groups/variables.tf:27]: Drop commented blocks

**Review by gsf — COMMENTED** [review]
The check for bcda-mgmt is failing (see https://github.com/CMSgov/ab2d-bcda-dpc-platform/actions/runs/9472959213/job/26099257878?pr=82) because the vpc module needs to account for that vpc having "managed" as the value for the "stack" tag. You should be able to add a line here to adjust for that: https://github.com/CMSgov/ab2d-bcda-dpc-platform/blob/3f2b09fb4bd5d0b40a0ee8d0ff7e52f472dd6bc4/terraform/modules/vpc/main.tf#L8

**Review by gsf — APPROVED** [review]
Looks good! Thanks for working through all the changes

**Pull request merged**